MiniApp/Classes/DownloadOperation.swift

```diff
@@ -0,0 +1,70 @@
+class DownloadOperation {
+    typealias DownloadCompletionHandler = (Result<URL, Error>) -> Void
+
+    private var completionHandler: DownloadCompletionHandler?
+    private var urlToDirectoryMap = [String: URL]()
+    private var miniAppClient: MiniAppClient
+    private var miniAppStorage: MiniAppStorage
+    private var miniAppPath: URL
+
+    private var queue: OperationQueue = {
+        let operationQueue = OperationQueue()
+        operationQueue.name = "MiniAppDownloader"
+        operationQueue.maxConcurrentOperationCount = 4
+        return operationQueue
+    }()
+
+    init(miniAppStorage: MiniAppStorage, miniAppClient: MiniAppClient, miniAppPath: URL, completionHandler: @escaping DownloadCompletionHandler) {
+        self.miniAppClient = miniAppClient
+        self.miniAppStorage = miniAppStorage
+        self.miniAppPath = miniAppPath
+        self.completionHandler = completionHandler
+    }
+
+    func downloadFiles(urls: [String]) {
+        self.miniAppClient.delegate = self
+        for url in urls {
+            guard let fileDirectory = UrlParser.parseForFileDirectory(with: url) else {
+                self.completionHandler?(.failure(NSError.downloadingFailed()))
+                return
+            }
+            if !FileManager.default.fileExists(atPath: self.miniAppPath.appendingPathComponent(fileDirectory).path) {
+                urlToDirectoryMap[url] = self.miniAppPath.appendingPathComponent(fileDirectory)
```

Comment:
Should this happen after a successful download of that file instead?

Reply:
We are also using this value capture to track the download progress that's why we are saving before the download starts

Reply:
Is it getting reset if the particular download fails?

Reply:
DownloadOperation object itself will be released once the download completes/fails